System.peers: enforce host_id

bhalevy opened the pull request:
The HOST_ID is already written to system.peers since inception pretty much (See https://github.com/scylladb/scylladb/pull/16376#discussion_r1429248185 for details).

However, it is written to the table using an individual CQL query and so it is not set atomically with other columns.
If scylla crashes or even hits an exception before updating the host_id, then system.peers might be left in an inconsistent state, and in particular without no HOST_ID value.

This series makes sure that HOST_ID is written to system.peers and use it to "seal" the record by upserting it in a single CQL BATCH query when adding the state for new nodes.

On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.

With that change we can assume that endpoint state loaded from system.peers will always have a valid host_id.

Refs https://github.com/scylladb/scylladb/pull/15903

gusev-p commented on service/storage_service.cc:
```diff
@@ -421,11 +421,11 @@ future<> storage_service::topology_state_load() {
                 co_await _sys_ks.local().update_peer_info(ip, "data_center", rs.datacenter);
                 co_await _sys_ks.local().update_peer_info(ip, "rack", rs.rack);
-                co_await _sys_ks.local().update_peer_info(ip, "host_id", id.uuid());
                 co_await _sys_ks.local().update_peer_info(ip, "release_version", rs.release_version);
```
Having many `INSERT`-s for a single row in `system.peers` seems wasteful. The `system.peers` table uses `schema_commitlog`, so each `INSERT` reaches the disk with `O_DSYNC`. We should probably
  * make a batch of info we want to write into `system.peers`;
  * compare it with the current state in the table;
  * write the changes in a single `INSERT`, if there are any. 

This is out of scope of this PR, deserving a separate one, WDYT @kbr-scylla

bhalevy replied:
makes sense to me and can be done as a follow up optimization

kbr-scylla replied:
> compare it with the current state in the table;

That would actually make it much more expensive, by introducing a read.

No need to do the comparison; we could keep doing unconditional write (but a single one instead of many).

Doesn't matter anyway. This is not a hot path.

gusev-p replied:
> That would actually make it much more expensive, by introducing a read.

Home come read (from cache, I suppose) can be more expensive than write with `O_DSYNC`?

> Doesn't matter anyway. This is not a hot path.

I wonder why, `O_DSYNC` write should be extremely expensive, IIUC.

kbr-scylla replied:
> Home come read (from cache, I suppose) can be more expensive than write with O_DSYNC?

Doesn't have to be from cache. If we ensured that it's cached then maybe.

> I wonder why, O_DSYNC write should be extremely expensive, IIUC.

By not a hot path I mean not used often. It only happens once per topology change. The hot paths are user data writes/reads which happen thousands times per second per shard.

gusev-p replied:
This is just unjustified pessimization, having them here and there makes the overall topology change procedure slow.

bhalevy replied:
done in v2

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5180/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6575/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5180/)
#### Build Details:
- Duration: 2 hr 16 min
- Builder: spider5.cloudius-systems.com

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -421,10 +421,9 @@ future<> storage_service::topology_state_load() {
+                co_await _sys_ks.local().update_tokens(ip, id.uuid(), rs.ring.value().tokens);
                 co_await _sys_ks.local().update_peer_info(ip, "data_center", rs.datacenter);
                 co_await _sys_ks.local().update_peer_info(ip, "rack", rs.rack);
-                co_await _sys_ks.local().update_peer_info(ip, "host_id", id.uuid());
```
This would also call update_cached_values. Which apparently does nothing. Maybe we should remove the function.

bhalevy replied:
sure

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -421,10 +421,11 @@ future<> storage_service::topology_state_load() {
                 co_await _sys_ks.local().update_peer_info(ip, "data_center", rs.datacenter);
                 co_await _sys_ks.local().update_peer_info(ip, "rack", rs.rack);
                 co_await _sys_ks.local().update_peer_info(ip, "release_version", rs.release_version);
+                // Finally, "seal" the state with host_id and tokens
```
Hmm... we call `update_peer_info(ip)` at the beginning of this sequence. So the row actually gets host ID written before that, at least if gossiper is populated when we do this. Which means we don't achieve the goal explained in the commit message :(

bhalevy replied:
I'll look into accumulating the changes and then applying them in a single cql query.

bhalevy replied:
done in v2

bhalevy commented:
In v2 ([de4a404](https://github.com/scylladb/scylladb/commit/de4a4042d0f7c66927d3a3294b67fb5517a9d08a)):
- batch update system.peers
  - well, that required some pretty-significant refactoring
  - but the end result is much cleaner IMO, so it's a net win both in correctness, performance, and code quality.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5272/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6676/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5272/)
#### Failed Tests (6/23439):
- [test_gms](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/test_misc/Tests___Unit_Tests___test_gms) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gms)
- [test_gms](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/test_misc/Tests___Unit_Tests___test_gms_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gms)
- [test_gms](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/test_misc/Tests___Unit_Tests___test_gms_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gms)
- [scylla-gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla-gdb.run.release.1)
- [scylla-gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla-gdb.run.release.2)
- [scylla-gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5272/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla-gdb.run.release.3)
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
In v3 ([1636d77](https://github.com/scylladb/scylladb/commit/1636d7717aad7047745c27cb8765a0fdb686db94)):
- Fix scylla-gdb to try accessing the application state map using std::unordered_map first, and fallback to std::map

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5275/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6680/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5275/)
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
> Make sure that HOST_ID is written to system.peers and use it to "seal" the record by updating it last (along with TOKENS) when adding the state for new nodes.
> 
> On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.

How does this work during upgrade?


> 
> With that change we can assume that endpoint state loaded from system.peers will always have a valid host_id.
> 
> Refs #15903

avikivity commented on cql3/query_processor.cc:
```diff
@@ -896,12 +903,23 @@ query_processor::execute_internal(
         db::consistency_level cl,
         service::query_state& query_state,
-        const std::initializer_list<data_value>& values) {
+        const Range& values) {
```
I think this could be done more cleanly by having a `data_value_list` type that's a wrapper over `std::vector<raw_value>`. Give that two non-explicit constructors - from `std::initializer_list<data_value>` and your range. Then execute_with_params() passes `data_value_list` to make_internal_options. This limits the templating to a small area.

bhalevy replied:
Sounds promising.  I'll try that

avikivity commented on types/types.hh:
```diff
@@ -1002,3 +1002,52 @@ struct appending_hash<data_type> {
+
+class data_values_map {
+    std::unordered_map<std::string, data_value> _map;
+
```
Why std::string and not sstring?

Why a wrapper over a map? Can use the map type directly.

bhalevy replied:
Why sstring?  I thought we're trying to stick with the standard when there's no special reason to use sstring (like the uninitialized construct case)

bhalevy replied:
> Why a wrapper over a map? Can use the map type directly.

To make the update case simpler.
With bare map `m["foo"] = "bar"` doesn't work because data_value can't be constructed this way.

avikivity replied:
unordered_map::insert_or_assign()

If you're solving a common problem, look for a common solution.

avikivity replied:
> Why sstring? I thought we're trying to stick with the standard when there's no special reason to use sstring (like the uninitialized construct case)

We're using sstring over std::string to avoid conversions when we do need sstring.

gusev-p replied:
I thought the main argument for `sstring` over `std::string` is performance, since the former doesn't have to be thread safe. Isn't this the case?

avikivity replied:
Recent (C++11) std::string is also not using copy-on-write. But it's still slower than sstring (sstring doesn't have a separate capacity word).

So we use sstring where performance is important, but also generally to avoid conversions.

avikivity commented on types/types.hh:
```diff
@@ -1002,3 +1002,52 @@ struct appending_hash<data_type> {
+            return true;
+        }
+        it->second = std::move(value);
+        return false;
```
Isn't this operator[]?

bhalevy replied:
I wish, but the compiler is playing hard to get (or actually hard to set in this case :))

avikivity replied:
I guess it's because the value is not default constructible.

There's insert_or_assign() though.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1668,6 +1668,37 @@ template future<> system_keyspace::update_peer_info<sstring>(gms::inet_address e
+
+    auto req = fmt::format("INSERT INTO system.{} ({}) VALUES ({})", PEERS, fmt::join(column_names, ","), values_ph_str);
+    slogger.debug("INSERT INTO system.{} ({}) VALUES ({})", PEERS, fmt::join(column_names, ","), fmt::join(data_values, ","));
+    co_await execute_cql_query(req, data_values).discard_result();
```
It's generally bad practice to generate CQL on the fly. CQL represents updates to structured data.

If you need to merge two updates, use BATCH.

bhalevy replied:
That would involve batchlog, wouldn't it?
The whole point of this series is to update system.peers with a single commitlog write.
Would we lose that if we use a batch insert?

avikivity replied:
No, single partition batch doesn't touch batchlog. It's just a way to generate a mutation.

avikivity commented on service/storage_service.cc:
```diff
@@ -4135,18 +4135,71 @@ future<> storage_service::do_update_system_peers_table(gms::inet_address endpoin
+            insert_string("supported_features", value);
+            break;
+        default:
+            break;
```
Should use BATCH, or the ugly UNSET feature thing.

avikivity commented on service/storage_service.cc:
```diff
@@ -429,13 +429,14 @@ future<> storage_service::topology_state_load() {
+                info.update("rack", rs.rack);
+                info.update("host_id", id.uuid());
+                info.update("release_version", rs.release_version);
+                co_await _sys_ks.local().update_peer_info(ip, std::move(info));
```
Could be a single fixed CQL statement (apart from merging the table name).

bhalevy replied:
true
We can pass CQL to uodate_peer_info rather than a column/value map.

avikivity replied:
Passing CQL strings is also not the cleanest.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1569,12 +1569,26 @@ data_value system_keyspace::make_tokens_data_value(const std::unordered_set<dht:
+
 future<std::unordered_map<gms::inet_address, std::unordered_set<dht::token>>> system_keyspace::load_tokens() {
-    sstring req = format("SELECT peer, tokens FROM system.{}", PEERS);
+    sstring req = format("SELECT peer, host_id, tokens FROM system.{}", PEERS);
```
How does this handle mixed clusters?

bhalevy replied:
All supported versions already store host_id in system.peers.  It shouldn't be a problem.

avikivity replied:
When did we start writing it?

bhalevy replied:
4311662828aa09627c77d9af16478f22624b6a25 called `update_peer_info` from `storage_service::handle_state_normal` back in `0.10`.
Note the use of `gossiper::uses_host_id` which was true from inception (See 6e83e54b0d61adc596a90e30d7ec4674d2176863 where I got rid of it)

The pretty-much initial implementation of `prepare_to_join` at efe067fd749626480c45ddf34de62a8853aa162c already populated the `HOST_ID` in the endpoint state, and it is sent in gossip also basically from inception (eed2795abe99ffc1619f7170707032e34777e350).

e86d39faf021f2797fdbbca93dfa9016e9c72430 moved the call to `update_peer_info` a bit later in `handle_state_normal`, when the node becomes a normal token owner (in branch 4.6)
but it doesn't matter since we've wrote the host_id to system.peers even before that.

avikivity replied:
ok, so no compatibility issues.

bhalevy commented:
> > Make sure that HOST_ID is written to system.peers and use it to "seal" the record by updating it last (along with TOKENS) when adding the state for new nodes.
> > On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.
> 
> How does this work during upgrade?

We're already populating the host_id in system.peers in the normal path.  This series just enforces it.

avikivity commented:
> > > Make sure that HOST_ID is written to system.peers and use it to "seal" the record by updating it last (along with TOKENS) when adding the state for new nodes.
> > > On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.
> > 
> > 
> > How does this work during upgrade?
> 
> We're already populating the host_id in system.peers in the normal path. This series just enforces it.

Since which version?

bhalevy commented:
> > > > Make sure that HOST_ID is written to system.peers and use it to "seal" the record by updating it last (along with TOKENS) when adding the state for new nodes.
> > > > On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.
> > > 
> > > 
> > > How does this work during upgrade?
> > 
> > 
> > We're already populating the host_id in system.peers in the normal path. This series just enforces it.
> 
> Since which version?

See https://github.com/scylladb/scylladb/pull/16376#discussion_r1429248185
Should be 0.X

avikivity commented:
> > > > > Make sure that HOST_ID is written to system.peers and use it to "seal" the record by updating it last (along with TOKENS) when adding the state for new nodes.
> > > > > On the read side, skip rows that have no HOST_ID state in system.peers, assuming they are incomplete, i.e. scylla got an exception or crashed while writing them, so they can't be trusted.
> > > > 
> > > > 
> > > > How does this work during upgrade?
> > > 
> > > 
> > > We're already populating the host_id in system.peers in the normal path. This series just enforces it.
> > 
> > 
> > Since which version?
> 
> See [#16376 (comment)](https://github.com/scylladb/scylladb/pull/16376#discussion_r1429248185) Should be 0.X

Please note it in the cover letter.

bhalevy commented:
In v4 ([0e20709](https://github.com/scylladb/scylladb/commit/0e207091fc4c7ddce0ec01a5083e1fb138396c67)):
- cql3: query_processor: add data_values_list
  - wrap `std::vector<data_value>` and provide implicit conversion from `initializer_list` and `std::ranges::range`
- system_keyspace: execute_cql_query: use data_values_list
  - no templates involved
- system_keyspace: add update_peer_info accepting a data_values_map
  - use cql batch
- gms: endpoint_state: define application_state_map
- storage_service: update_peer_info: pass a data_values_map to sys_ks
  - refactored `get_peer_info_for_update`
- system_keyspace: update_peer_info: drop single-column overload

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5285/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6696/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5285/)
#### Build Details:
- Duration: 2 hr 27 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
- rebased
- updated cover letter

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5439/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5439/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6852/)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider3.cloudius-systems.com

avikivity commented on cql3/query_processor.hh:
```diff
@@ -341,9 +347,13 @@ public:
+        co_return co_await execute_internal(query_string, cl, qs, values, cache);
+    }
+
     future<::shared_ptr<untyped_result_set>> execute_internal(
```
I think this patch (cql3: query_processor: define execute_internal overload in header) is no longer needed.

bhalevy replied:
right

avikivity commented on types/types.hh:
```diff
@@ -1002,3 +1004,14 @@ struct appending_hash<data_type> {
     }
 };
+
+class data_values_list : public std::vector<data_value> {
```
data_value_list is more grammatically accurate.

bhalevy replied:
okay.  I see you used `data_values_vector` below in your example.
Which do you prefer?  (Both look good to me)

avikivity replied:
I meant to write data_values_list to be consistent with what you wrote. 

Either data_value_vector or data_value_list are fine.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1668,6 +1668,34 @@ template future<> system_keyspace::update_peer_info<sstring>(gms::inet_address e
+
+    co_await execute_cql_query(ss.str(), values).discard_result();
+}
+
```
I still don't like it. We have a fixed schema for the table, and a system for using prepared statements, and it gets thrown away.

We can use either a batch or prepared statements, or the UNSET thing, but not this.

We can have something like

```c++
struct batch_element {
    statements::prepared_statement::checked_weak_ptr stmt; // from prepare_internal   
    data_values_vector arguments;
};

future<> query_processor::execute_internal_batch(const std::vector<batch_element>);
```

bhalevy replied:
So with UNSET we can prepare a INSERT/UPDATE query specifying all columns and provide UNSET values for the values that are unavailable?

bhalevy replied:
Unless I misunderstood something using an unset value seems very complicated and requires highly intrusive changes.

avikivity replied:
Why is unset difficult? You can make data_values_list accept two constructors, one from convertible_to\<data_value>, and one from variant\<data_value, unset>. The first will be used by practically everyone. The second will be used by those crazies that want to use unset.

It's true the callers will have to collect their inputs in some struct and provide it later, but it's localized to those callers. For the peers table:

```c++
struct peer_optional_values {
    std::optional<sstring> data_center;
    ...
};
```

and then the UPDATE statement will pick up values from peer_optional_values.

bhalevy replied:
Maybe I don't understand how it's supposed to work but after exploring this it looks like unset is not being handled on the modification path, only `null` is handled, and the latter deletes the cells. I guess it is only handled for the WHERE clause expressions.

gusev-p replied:
can't we just have a struct with all the fields we want to update in `update_peer_info`
```
struct peer_info_update_request {
  std::optional<inet_address> rpc_address;
  std::optional<sstring> rack;
  // ...
}
```
`update_peer_info` does `select * from peers`, merges the data with the provided `peer_info_update_request` and inserts them back with a single big static insert:
  * it statically describes the interface of `update_peer_info`, we actually don't need dynamism in this dimension;
  * we don't need all the complexities with `data_values_list`, as long as its overheads (allocation for the vector/`data_value` copies) for the data plane;
  * we can skip the insert altogether if all the values from `peer_info_update_request` are equal to the current ones;
  * simple insert allows for prepared statements

bhalevy commented on types/types.hh:
```diff
@@ -1002,3 +1004,14 @@ struct appending_hash<data_type> {
     }
 };
+
+class data_values_list : public std::vector<data_value> {
```
I wouldn't worry about it too much since it should be a rare event, but it makes sense.

gusev-p replied:
IIUC, in [this commit](https://github.com/scylladb/scylladb/pull/16376/commits/0657e4ed187a0782d4655bb02534ede301a25d6c) we add these allocations and copies to all the call sites of the methods with `data_values_list` parameter which uses `initializer_list`.

bhalevy replied:
fair point

bhalevy commented:
> can't we just have a struct with all the fields we want to update in update_peer_info
>
> struct peer_info_update_request {
>   std::optional<inet_address> rpc_address;
>   std::optional<sstring> rack;
>   // ...
> }
> update_peer_info does select * from peers, merges the data with the provided peer_info_update_request and inserts them back with a single big static insert:

First, this will require read before write.
And what do you mean by static insert?
An insert query for all columns?

>
> it statically describes the interface of update_peer_info, we actually don't need dynamism in this dimension;
we don't need all the complexities with data_values_list, as long as its overheads (allocation for the vector/data_value copies) for the data plane;

I worry more about the complexity than the allocation (it's only 9 columns tops)

> we can skip the insert altogether if all the values from peer_info_update_request are equal to the current ones;
simple insert allows for prepared statements

That's true. Yet requires read before right (and ideally, locking, since it's not atomic)

bhalevy commented:
In addition to the above, even with read before write, we still need to deal with null columns before they are set the first time.  But in this case it wouldn't hurt passing null values for missing columns as this will just preserves their noon-existence.

gusev-p commented:
> First, this will require read before write.

Yes. I thought it should be quite cheaper. Maybe we can get rid of it with `UNSET_VALUE` for all the fields not mentioned in `peer_info_update_request`.

> And what do you mean by static insert?

We don't need to construct the text of the query dynamically.

bhalevy commented:
> > First, this will require read before write.
> 
> Yes. I thought it should be quite cheaper. Maybe we can get rid of it with `UNSET_VALUE` for all the fields not mentioned in `peer_info_update_request`.
> 
> > And what do you mean by static insert?
> 
> We don't need to construct the text of the query dynamically.

I believe that the problem is not really with constructing the text dynamically, but rather ending up with ad-hoc queries that stress query preparation.

gusev-p commented:
> I believe that the problem is not really with constructing the text dynamically, but rather ending up with ad-hoc queries that stress query preparation.

This is what I actually meant here.

bhalevy commented:
In v5 ([7a53397](https://github.com/scylladb/scylladb/commit/7a5339701bc36174762d8c243b1c11f308dfddf8))
- rebased
- defined struct peer_info with optional members for each system.peer column
- use `std::vector<std::optional<data_value>>` to pass values to query_processor with nullopt standing for unset values
  - query processor translates it to `cql3::raw_value_vector_with_unset` for preparing the query options
- dropped a few, now-unneeded patches

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5467/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5467/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6884/)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
[a6cff77](https://github.com/scylladb/scylladb/commit/a6cff77998ac6c91c4e9870844e8269f61d0c338): changed inadvertent logger.info call to logger.debug

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5472/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5472/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6888/)
#### Failed Tests (1/23683):
- [test_nulls_in_user_types](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5472/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_nulls_in_user_types) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nulls_in_user_types)
#### Build Details:
- Duration: 1 hr 41 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
> > > First, this will require read before write.
> > 
> > 
> > Yes. I thought it should be quite cheaper. Maybe we can get rid of it with `UNSET_VALUE` for all the fields not mentioned in `peer_info_update_request`.
> > > And what do you mean by static insert?
> > 
> > 
> > We don't need to construct the text of the query dynamically.
> 
> I believe that the problem is not really with constructing the text dynamically, but rather ending up with ad-hoc queries that stress query preparation.

I'm not worried about stress, since this occurs very infrequently, but about violating principles of interacting with database (I guess those were motivated by stress, but not just that).

avikivity commented on cql3/query_processor.cc:
```diff
@@ -794,6 +794,49 @@ query_options query_processor::make_internal_options(
 
+query_options query_processor::make_internal_options(
+        const statements::prepared_statement::checked_weak_ptr& p,
+        const std::vector<std::optional<data_value>>& opt_values,
```
We shouldn't use optional for unset. optional is very common, and is already used for null. Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.

That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).

avikivity replied:
Also,  why is this a new overload?

bhalevy replied:
> We shouldn't use optional for unset. optional is very common, and is already used for null.

Used for null `data_value`?  I though that's what `data_value::make_null()` is for.


> Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.
> 
> That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).

My first attempt (that I didn't publish) was to extend data_value to hold an unset value in addition to a null value, but that was overly complicated and intrusive.

I realized that a null value is considered a data_value while a unset value is not a value.  It's a lack of thereof.
That's why a disengaged `optional<data_value>` should mean "no value" rather than "null value".

But I can live with the variant as well.

> Also, why is this a new overload?

To save an extra allocation to store the initializer_list into a vector in the common case.

avikivity replied:
> > We shouldn't use optional for unset. optional is very common, and is already used for null.
> 
> Used for null `data_value`? I though that's what `data_value::make_null()` is for.


Not for null data_value, but for null managed_bytes_opt. At the caller site it's easy to mistake one of the other.

In any case std::nullopt doesn't hint at UNSET, which is an obscure feature, so no one will think of it.

> 
> > Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.
> > That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).
> 
> My first attempt (that I didn't publish) was to extend data_value to hold an unset value in addition to a null value, but that was overly complicated and intrusive.


Yes. It's also inserting an obscure feature into a main API.

> 
> I realized that a null value is considered a data_value while a unset value is not a value. It's a lack of thereof. That's why a disengaged `optional<data_value>` should mean "no value" rather than "null value".


It's too subtle. People are used to seeing std::nullopt as null due to managed_bytes_opt.

> 
> But I can live with the variant as well.
> 
> > Also, why is this a new overload?
> 
> To save an extra allocation to store the initializer_list into a vector in the common case.


All of this is a slow path for metadata. The data path builds mutations directly.

initializer_list<> is a huge pessimiziation since it can't be moved from.

avikivity commented on cql3/query_processor.hh:
```diff
@@ -377,6 +377,25 @@ public:
+            service::query_state& query_state,
+            const std::vector<std::optional<data_value>>&);
+
+
```
Too many new overloads. Let's make the existing ones work. The old data_values_list could have been taught to construct itself from unset markers.

bhalevy replied:
Can do.  But I'll base it on `small_vector` based on https://github.com/scylladb/scylladb/pull/16376#discussion_r1438084234

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1653,20 +1650,116 @@ future<std::unordered_map<gms::inet_address, gms::inet_address>> system_keyspace
+    } else {
+        on_fatal_internal_error(slogger, fmt::format("unsupported sstring peer info column_name \"{}\"", column_name));
+    }
+    co_await update_peer_info(ep, info);
```
This is really odd, one update_peer_info calling another, but it's not recursion. This is just trying to confuse readers.

bhalevy replied:
As I wrote in the comment above, it's just a temporary crutches to reduce code churn and introduce the changes one layer at a time to aid reviewability and testibility of the series.

avikivity replied:
Well it would have been better to use a different name.

bhalevy replied:
It's a pre-existing overload (see https://github.com/scylladb/scylladb/blob/f1dea4bc8ad8b443733454aa5abb319c4d1da230/db/system_keyspace.cc#L1661-L1675)
and it's going away in this series, so I don't think it's worth the trouble.

avikivity replied:
ok

gusev-p commented on gms/endpoint_state.hh:
```diff
@@ -19,7 +19,7 @@
 namespace gms {
 
-using application_state_map = std::map<application_state, versioned_value>;
+using application_state_map = std::unordered_map<application_state, versioned_value>;
```
Why not `absl::flat_hash_map`? It's way more efficient than `std::unordered_map`?

I wonder why it's still not the default choice for dictionary data structure in the Scylla code.

gusev-p replied:
Great series, a lot of cleanups, thanks!

bhalevy replied:
> Why not `absl::flat_hash_map`? It's way more efficient than `std::unordered_map`?
> 
> I wonder why it's still not the default choice for dictionary data structure in the Scylla code.

I don't know if idl supports it.

> Great series, a lot of cleanups, thanks!

Thanks.

bhalevy commented:
> > I realized that a null value is considered a data_value while a unset value is not a value. It's a lack of thereof.
> >That's why a disengaged optional<data_value> should mean "no value" rather than "null value".
> >
> I like this, variant<data_value, unset_value> is also ok.
>
> > To save an extra allocation to store the initializer_list into a vector in the common case.
>
> Why we need a vector for std::optional-s? Can't we just turn the existing overloads into > std::initializer_list<std::optional<data_value>>? I think there is too much copypaste here, it will make the code harder to maintain.
>
> I think we can construct initializer_list just like we construct vector in update_peer_info:
> ```
> bool any_set = false;
> std::initializer_list<std::optional<data_value>> opt_values {
>   to_opt_value(info.data_center, &any_set),
>   to_opt_value(info.host_id, &any_set),
>   ...
> };
> ```

In the next version I defined:
```
using data_value_or_unset = std::variant<data_value, unset_value>;
class data_value_vector : public utils::small_vector<data_value_or_unset, default_data_value_vector_size>
```

It reduces the churn and is more flexible.
It is constructible either from initializer_list<data_value> or from initializer_list<data_value_or_unset>.

bhalevy commented:
In v6 ([cdd5605](https://github.com/scylladb/scylladb/commit/cdd5605d812929fb41f900ce5768f2db39aa80eb)):
- `using data_value_or_unset = std::variant<data_value, unset_value>;`
- `using data_value_list = std::initializer_list<data_value_or_unset>;`

avikivity commented on cql3/query_processor.cc:
```diff
@@ -771,27 +771,37 @@ query_options query_processor::make_internal_options(
+        std::visit(overloaded_functor {
+            [&] (const data_value& v) {
+                if (v.type() == bytes_type) {
+                    bound_values.values.emplace_back(cql3::raw_value::make_value(value_cast<bytes>(v)));
```
Note: this special case is probably a pessimization, since it forces linearization. It's not critical since I don't expect huge values here, but it would be nice to remove it, separately.

avikivity commented on types/types.hh:
```diff
@@ -1004,4 +1005,33 @@ struct appending_hash<data_type> {
+    }
+};
+
+using data_value_list = std::initializer_list<data_value_or_unset>;
```
I'm impressed the compiler is able to implicitly convert through so many layers.

avikivity commented on types/types.hh:
```diff
@@ -1004,4 +1005,33 @@ struct appending_hash<data_type> {
+            },
+            [&ctx] (const unset_value& u) {
+                return fmt::format_to(ctx.out(), "{}", u);
+            }
```
These two could have been merged ([&] (const auto& v)  { ... }). But no matter.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5480/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6900/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5480/)
#### Build Details:
- Duration: 1 hr 40 min
- Builder: spider5.cloudius-systems.com

gusev-p commented on cql3/query_processor.cc:
```diff
@@ -763,7 +763,7 @@ std::pair<std::reference_wrapper<struct query_processor::remote>, gate::holder>
 
 query_options query_processor::make_internal_options(
```
nit: commit message:
> data_value_list is a wrapper around std::initializer_list<data_value>.

doesn't correspond to the code (`data_value_list` is defined to be equal to `std::initializer_list<data_value>;` in this commit)

gusev-p commented on cql3/query_processor.cc:
```diff
@@ -771,27 +771,37 @@ query_options query_processor::make_internal_options(
                 format("Invalid number of values. Expecting {:d} but got {:d}", p->bound_names.size(), values.size()));
     }
```
nit: commit message:
> Add overloads for execute_internal and friends
> accepting a vector of optional<data_value>.

doesn't correspond to the code

gusev-p commented on service/storage_service.cc:
```diff
@@ -485,8 +486,9 @@ future<> storage_service::topology_state_load() {
                 if (rs.ring.has_value()) {
                     if (!is_me(ip)) {
                         // Save ip -> id mapping in peers table because we need it on restart, but do not save tokens until owned
-                        co_await _sys_ks.local().update_tokens(ip, {});
```
we lost `co_await _sys_ks.local().update_tokens(ip, {});` here. I'm not sure if it's important.

bhalevy replied:
if no tokens are stored yet' then storing an empty set is equivalent (yet I think it's better to store no tokens than to store an empty set).
If there were tokens than storing an empty set here is definitely wrong.  The mode should be decommissioned or removed to lose its tokens.

bhalevy commented:
> It's a matter of just copy-pasting (or templatizing) this [std::map](https://github.com/scylladb/scylladb/blob/246da8884a519ba886202dcce79de15440c7fc60/serializer_impl.hh#L373) serializer. There is already absl::btree_set support there.
> 
> Note: there is no compatibility issue here since std::map and std::unordered_map are serialised/deserialized in the same way. The same could be true for `absl::flat_hash_map`

It's really out of scope for this PR.
I regret even including the change to unordered_set.
Let's drop it and the we can make the change to abseil in one go.
You're invited to craft a patch for it.
